### PR TITLE
Bug Bounty: NPSP RD Settings Apex Class enforce Salesforce Security

### PR DESCRIPTION
**Pull request by andrewyu-salesforce**
NPSP Settings should follow Salesforce security best practices at every level.
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by andrewyu-salesforce** [comment]
**lurch: attache W-041508

**Comment by andrewyu-salesforce** [comment]
@force2b The PR is ready for review
@sam-knox Please review the error message stored onto the Error record. The error message is reused from the Field Level Security Error message that already exist in Custom Label
Screen Shot 2020-06-11 at 3 40 43 PM

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007haS9QAI" target="_blank">W-041508</a>

File: src/classes/STG_PanelRD_CTRL.cls [hunk 1]
```diff
@@ -81,10 +81,21 @@ public with sharing class STG_PanelRD_CTRL extends STG_Panel {
         Savepoint sp = Database.setSavepoint();
         try {
         	List<SObject> toUpdate = new List<SObject>();
-
         	if (STG_Panel.stgService.stgRD != null && !Test.isRunningTest()) {
         		toUpdate.add(STG_Panel.stgService.stgRD);
-        	}
+            }
```
Comment (force2b) [inline src/classes/STG_PanelRD_CTRL.cls:86]: The indentation looks off here. I suspect that there may be some tabs in this class versus spaces. You should be able to convert the indentation to spaces for the entire class.

**Comment by andrewyu-salesforce** [comment]
@force2b Thanks for the review. The tabs are converted to space.

**Comment by rponti-sforg** [comment]
@andrewyu-salesforce @force2b  I am not sure if Feature Test and Feature Test - Robot have actually stalled or if they were restarted.

**Comment by rponti-sforg** [comment]
> @andrewyu-salesforce @force2b I am not sure if Feature Test and Feature Test - Robot have actually stalled or if they were restarted.

Scratch that ... I see Mike restarted the Feature Test.  Thank you.

**Comment by sam-knox** [comment]
@andrewyu-salesforce - Yes, the error message for trying to access RD settings is fine. Another question though: in this PR are we introducing any more security checks? We already check that a user has Customize Application, Modify All Data, and Author Apex. Are we expanding those checks at all here?

**Comment by andrewyu-salesforce** [comment]
@sam-knox This does not introduce any new security check. This PR just add an extra security to check in the back end to ensure the user who bypass our UI security still need required permission ('Customize Application') to execute the code.

**Comment by rponti-sforg** [comment]
> @andrewyu-salesforce - Yes, the error message for trying to access RD settings is fine. Another question though: in this PR are we introducing any more security checks? We already check that a user has Customize Application, Modify All Data, and Author Apex. Are we expanding those checks at all here?

@sam-knox I had a similar question.  For this PR / WI, we are focused on the '**Customize Application**' permission.  So, I think the only _security check_ being made here is on the '**Customize Application**' permission.  

This was Andrew's response to a similar question I had:

"_That's correct. Custom Settings does not respect Modify All permission at all. It's purely controlled by Customize Application permission. Also, Without Author Apex permission, Users cannot call the saveSettings() function.

This security bug only occurs when you go through Dev Console Script(System Mode). If you go through UI (User Mode), Core will automatically handle the error , and the error message will be different. I believe the UI part does not needed to be tested because it's default Salesforce behavior._"

@andrewyu-salesforce  Please let us know if we if we are missing something.

**Comment by rponti-sforg** [comment]
QE testing for this PR and associated WI is now complete.  Based on the WI Description, Developer feedback, GUS WI, and Test Charter, tests have passed.

For further testing details, see:
https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000007haS9QAI/view
https://salesforce.quip.com/iVMVAzIRQVCf

Proceeding with merging PR and setting WI status to 'Closed' once the post merge builds have passed.

**Pull request merged**